### DISCUSSION TITLE
React wrapper

### DISCUSSION THREAD

**Discussion by yplaxma** [MathKeyboardEngine/MathKeyboardEngine#2] [Q&A] [answered]
Hello. First of all absolutely great project, I love the focus and elegance of doing one particular thing well.
I'd like to use the package as it fits my use-case perfectly but I'm using React and not used to non-react components.
Are there any plans to make a React wrapper?

**Comment by SymboLinker** [comment]
Hi!

> Hello. First of all absolutely great project, I love the focus and elegance of doing one particular thing well.

 Thanks for the compliment :)

> I'd like to use the package as it fits my use-case perfectly but I'm using React and not used to non-react components.
Are there any plans to make a React wrapper?

I guess that making a wrapper is not needed. It seems that when using React, you can use npm. I will try it out this week (I have never used React before) and I will provide you with steps to use MathKeyboardEngine in React.

**Comment by SymboLinker** [comment, marked as answer]
**_Steps to use mathkeyboardengine in React: a sample app tutorial_**

### Part 1 - Create a React App
1. Install [Node.js](]https://nodejs.org/).
2. Install [VS Code](https://code.visualstudio.com/) (used in this tutorial, but you can use another code editor).
3. Create a folder "my-react-test" and open that folder in VS Code.
4. Click `Terminal` > `New Terminal` and execute the following commands (taken from [reactjs.org](https://reactjs.org/docs/create-a-new-react-app.html#create-react-app)):
```
npx create-react-app my-app
cd my-app
npm start
```
(If npx or npm is not found, restart VS Code and try again. You may have installed Node.js after installing VS Code.)

5. A browser window with URL `http://localhost:3000/` is opened. Among other things, you see the sentence `"Edit src/App.js and save to reload."`.

.
### Part 2 - Add MathKeyboardEngine to the newly created React App
1. Click `Terminal` > `New Terminal` and execute the following command:
```
npm i mathkeyboardengine
```
2. Open `src/App.js` and replace its contents by:
```
import './App.css';
import * as mke from 'mathkeyboardengine';

let config = new mke.LatexConfiguration();
let k = new mke.KeyboardMemory();
mke.insert(k, new mke.DigitNode('1'));
mke.insert(k, new mke.DigitNode('2'));
mke.insertWithEncapsulateCurrent(k, new mke.DescendingBranchingNode(String.raw`\frac{`, '}{', '}'));
let editModeLatex = mke.getEditModeLatex(k, config);

function App() {
  return (
    <div className="App">
      <header className="App-header">
        <h1>{editModeLatex}</h1>
        <p>
          Edit <code>src/App.js</code> and save to reload.
        </p>
      </header>
    </div>
  );
}

export default App;
```
3. The result:
![afbeelding](https://user-images.githubusercontent.com/32139898/193336447-17d5c716-eab8-4242-b495-caecd5ce66a5.png)
Conclusion: the existing npm package can be used in React apps.

.
### Part 3 - Add react-katex to the React App that uses MathKeyboardEngine
This part is optional: the previous part "proves" that MathKeyboardEngine can be used in React Apps. However, very probably you will also use a typesetting library and it may be nice to have an example of that too... for this tutorial I picked [react-katex](https://www.npmjs.com/package/react-katex).
1. Click `Terminal` > `New Terminal` and execute the following command:
```
npm i react-katex
```
2. Now add the following imports:
```
import 'katex/dist/katex.min.css';
import { InlineMath, BlockMath } from 'react-katex';
```
3. And use one of them:
```
<BlockMath math={editModeLatex}/>
```
4. The result:
![afbeelding](https://user-images.githubusercontent.com/32139898/193338322-6757b08f-24c2-4bdd-aa89-0a841115ffd3.png)

### Part 4 - Customize!
Your app will probably not be a single-page app, so the import statements should be moved to other locations. Also, using `insert` will very probably depend on a click event or something like that, etc.

If you need any help, let me know.

**Reply by yplaxma** [reply]
Thank you for the fast and great reply! I had to focus on something else so couldn't check nor respond.

Before delving into the subject, I wanted to compliment you on how well organized everything within the project is. After starting this discussion I went ahead and tried to go through and understand the code and was surprised on how easy and seamless I was able to do that and it was all thanks to how neat everything is. I don't mean to be overly complimentary but it really was. This was a nice surprise because it's usually rough for me and takes a long time to understand how packages work since they're mostly geared towards being understood by the maintainers; not to mention I'm still on the beginner to intermediate level.

For React, I actually was able to eventually make it work. I didn't use react-katex because I use React 18 and they use an older version and I didn't want to deal with that, I did however use the same approach as they do (dangerouslySetInnerHtml).
The only issue I ended up with is how MKE's functions like insert are not particularly compatible with how React handles state since they mutate keyboardMemory directly instead of returning a new object. Not sure how much you read up on React but if you mutate state directly, React doesn't detect that the state changed so the component reflecting the keyboardMemory doesn't refresh to render the new value. This means that whenever we update keyboardMemory we actually have to return a new object instead of mutating the old one, if we want the change to be detected and re-rendered in React.

Of course there is always the "force refresh" workaround but it goes against the right way to do things. I think this could be fixed just by having any keyboardMemory update be followed by cloning the updated keyboardMemory and returning the clone. I personally like this approach because it would avoid too much diversion of the "React version" of the package from the main one. 
The only change in the main would be a way to clone keyboardMemory, this I'm not quite sure how to go about since I don't know what the right approach would be; maybe simply adding a constructor to the KeyboardMemory class which takes the KeyboardMemory object to be cloned?

**Reply by SymboLinker** [reply]
That is a kind reply - I am very happy with the words "seamless", "neat" and other things you mention :)

About the component updates in React: if there is no difference in performance and user experience, I would choose to write code in a way that expresses its intention best. If the intention is to force an update of a component, then just call `forceUpdate` for that component. Using the other way - making a clone of an object - would force people reading the code: (1) to verify that the clone is done correctly and (2) to do an additional thinking step "setting this clone will force an update". Having "hidden logic" like that - instead of being "code readable in a glance" - seems not clean. (I use the word "clean" here, referring to the concept "clean code" that is explained in the classic book "Clean code" by Robert C. Martin.)

Choose one way and use that consistently in your repository in a way that you can easily change. You can always change it if someone convinces you that something else is better.
